common/go-utils/src/inv.rs

```diff
@@ -103,20 +87,18 @@ pub fn list_upstream_artifacts() -> Result<Vec<Artifact>, String> {
         .map_err(|e| e.to_string())?
         .iter()
         .flat_map(|release| &release.files)
-        .filter(|file| !file.sha256.is_empty())
-        .filter_map(|file| {
-            file.target_arch()
-                .filter(|target| target == ARCH)
-                .map(|target| {
-                    Version::parse_go(&file.version)
-                        .map(|version| Artifact {
-                            go_version: file.version.clone(),
-                            semantic_version: version,
-                            architecture: target,
-                            sha_checksum: file.sha256.clone(),
-                        })
-                        .map_err(|e| e.to_string())
+        .filter(|file| !file.sha256.is_empty() && file.os == "linux" && file.arch == "amd64")
+        .map(|file| {
+            Version::parse_go(&file.version)
+                .map(|version| Artifact {
+                    go_version: file.version.clone(),
+                    semantic_version: version,
+                    os: String::from("linux"),
+                    arch: String::from("x86_64"),
+                    sha_checksum: file.sha256.clone(),
+                    url: format!("{}/{}", GO_HOST_URL, file.filename),
                 })
+                .map_err(|e| e.to_string())
```

Comment:
instead of coercing all errors into `String` values you should consider using an enum here.  E.g.;

```rust
#[derive(Debug, thiserror::Error)]
pub enum ListUpstreamArtifactsError {
    #[error(transparent)]
    InvalidResponse(ureq::Error),
    #[error(transparent)]
    ParseJsonResponse(std::io::Error),
    #[error(transparent)]
    InvalidArtifactVersion(VersionParseError)
}
```

and mapping all the errors that might happen during the execution of `list_upstream_artifacts` into this error type which might look like this:

```rust
pub fn list_upstream_artifacts() -> Result<Vec<Artifact>, ListUpstreamArtifactsError> {
    ureq::get(GO_RELEASES_URL)
        .call()
        .map_err(ListUpstreamArtifactsError::InvalidResponse)?
        .into_json::<Vec<GoRelease>>()
        .map_err(ListUpstreamArtifactsError::ParseJsonResponse)?
        .iter()
        .flat_map(|release| &release.files)
        .filter(|file| !file.sha256.is_empty() && file.os == "linux" && file.arch == "amd64")
        .map(|file| {
            Artifact::try_from(file).map_err(ListUpstreamArtifactsError::InvalidArtifactVersion)
        })
        .collect::<Result<Vec<_>, _>>()
}
```

the use of `#[error(transparent)]` from `thiserror::Error` for each error type delegates to the `Display` implementation of the wrapped error which is essentially the same as your `.map_err(|e| e.to_string())` calls but you could provide a custom error message for each enum value if you wanted.

Reply:
I agree, and this is on my list already -- but maybe I can handle that in a separate PR that does just that (as this one doesn't touch the error handling in this area)?

Reply:
sure. i wasn't going to block this PR for any of these suggestions. just pointing out Rust-ish things that i didn't find intuitive when i first started that you might also find helpful. feel free to do this later on 😄

Reply:
Ended up taking the approach you suggested (almost verbatim) anyway :) https://github.com/heroku/buildpacks-go/pull/229/commits/84b811c0b1896bb2c5daf2fd2d5cef57c15c3dbb